Suppress trim analysis warnings inside reflection implementation

vitek-karas opened the pull request:
See the Justification for each of the added suppressions.

vitek-karas committed:
Suppress trim analysis warnings inside reflection implementation

See the Justification for each of the added suppressions.

MichalStrehovsky commented on src/coreclr/src/System.Private.CoreLib/src/System/Reflection/RuntimeExceptionHandlingClause.cs:
```diff
@@ -38,6 +39,11 @@ public override int FilterOffset
 
         public override Type? CatchType
         {
+            [UnconditionalSuppressMessage("ReflectionAnalysis", "IL2026:RequiresUnreferencedCode",
```
Should we mark `MethodBody` -> `ExceptionHandlingCaluse` -> ... and friends as `RequiresUnreferencedCode` instead?

My line of thinking is that we want to warn whenever trimming can produce different results than pre-trimming. Since linker is going to potentially stub out method bodies, replace basic blocks with nops, etc. maybe these should all warn?

vitek-karas replied:
It seem marking `MethodBase.GetMethodBody()` with `RequiresUnreferencedCode` would be enough - it should cover changes to IL, local variables, exception clauses, ... As far as I can tell all of those have to start with somehow getting the `MethodBody` object, which seems to be only possible via `MethodBase.GetMethodBody`.

MichalStrehovsky reviewed (approved):
Looks good otherwise

vitek-karas committed:
Annotate GetMethodBody as potentially dangerous with trimming.

Trimming can change IL of a method (can remove branches, remove local variables, change some isntructions, ...) - as such accessing the actual body of a method is potentially dangerous.

vitek-karas committed:
Fix unintended change

vitek-karas committed:
Fix order of usings

vitek-karas committed:
Update ref assembly with the new attribute

vitek-karas committed:
Propagate the new annotation to Mono's CoreLib as well

vitek-karas committed:
Fix the standalone build of EventSource
